_posts/2021-09-03-rudiments-of-data-wrangling-in-rust.md

```diff
@@ -0,0 +1,519 @@
+how quickly EVs have become more popular here in the recent years.
+
+![Electric vehicles](/img/rudiments-of-data-wrangling-in-rust/electric-vehicles.jpg)
+
+*Photo by <a href="https://unsplash.com/@michaelfousert?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText" target="_blank">Michael Fousert</a> on <a href="https://unsplash.com/s/photos/electric-vehicles?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText" target="_blank">Unsplash</a>*
+
+[Traficom](https://www.traficom.fi/en/) (Finnish Transport and Communications Agency) regularly
+publishes information about registered vehicles in active use, 
+as downloadable [open data](https://www.traficom.fi/en/news/open-data?toggle=Open%20data%20for%20vehicles%20version%205.14). This data includes the driving power of 
+the vehicle, its type and first registration date, and many other
+data points.
+
+The data is in CSV format, and it is accompanied by a Microsoft Excel
+spreadsheet with descriptions of the data fields. The latest iteration
+of the CSV file, published in March 2021, has over five million rows,
+and it includes all land vehicles (cars, buses, tractors etc.)
+
+## Reading CSV files in Rust
+
+The `csv` crate contains structs with functions to read and write CSV
+files, although we only use the reading part in this program.
+
+The CSV data file we use requires some special handling. Instead of
+a comma, it uses the semicolon as the field delimiter. Also, the text
+in the file is encoded in ISO 8859-1. The default settings of the 
+CSV reader in the csv crate expect comma-delimited files with text
+encoded in UTF-8 (which is the internal representation of Rust strings).
+
+To convert the lines in the CSV file from ISO 8859-1 to UTF-8 we use the
+`encoding_rs` and `encoding_rs_io` crates, which implement 
+the [Encoding Standard](https://encoding.spec.whatwg.org). 
+In their documentation you can also find a link to 
+an interesting long-form article which also tells you why ISO 8859-1 
+is treated as Windows code page 1252.
+(This conversion would not be needed if the source data were already 
+in UTF-8, like it should be.)
+
+In this case it means that we need one extra step to construct 
+our CSV reader from a transcoding reader supplied by the `encoding_rs_io`
+crate:
+
+    let file = File::open(filename)?;
+
+    let transcoded_reader = DecodeReaderBytesBuilder::new()
+        .encoding(Some(WINDOWS_1252))
+        .build(file);
+
+    let mut reader = csv::ReaderBuilder::new()
+        .has_headers(true)
+        .delimiter(b';')
+        .from_reader(transcoded_reader);
+
+We also told `csv::ReaderBuilder` that the first line of our CSV
+file is the header, and the field delimiter is the semicolon
+character.
+
+## Processing the CSV records
+
+Once we have our reader, we can start preparing to read the lines.
+From the reader we get an iterator which returns values of type
+`Result<csv::StringRecord, std::error::Error>`. Each `StringRecord`
```

Comment:
Would be nice to see some summary of how the Result types are a good example of why algebraic data types are useful. Also how they are what the typical error handling convention is built around.

I am not sure where this would actually fir or if this kind of structural change at this time is worth the benefit, but thought to mention it anyway

Reply:
I think that algebraic data types and their use for error handling are orthogonal. It would be good to be able to use a lot more bandwidth than is feasible here to introduce algebraic data types properly, with a motivating example, and without error handling at first, and then point out separately that they are well suited and heavily used for error handling in Rust. Maybe suitable material for another blog post?